diff --git a/python/google/protobuf/__init__.py b/python/google/protobuf/__init__.py
index 45a6c20c5..c28dd8852 100755
--- a/python/google/protobuf/__init__.py
+++ b/python/google/protobuf/__init__.py
@@ -8,3 +8,9 @@
 # Copyright 2007 Google Inc. All Rights Reserved.

 __version__ = '6.31.1'
+
+if __name__ != '__main__':
+  try:
+    __import__('pkg_resources').declare_namespace(__name__)
+  except ImportError:
+    __path__ = __import__('pkgutil').extend_path(__path__, __name__)

diff --git a/python/dist/system_python.bzl b/python/dist/system_python.bzl
index 390bf3c2a..2a1c56483 100644
--- a/python/dist/system_python.bzl
+++ b/python/dist/system_python.bzl
@@ -69,9 +69,11 @@ def fuzzing_py_install_deps():
 """
 
 _build_file = """
+load("@rules_shell//shell:sh_binary.bzl", "sh_binary")
 load("@bazel_skylib//lib:selects.bzl", "selects")
 load("@bazel_skylib//rules:common_settings.bzl", "string_flag")
-load("@bazel_tools//tools/python:toolchain.bzl", "py_runtime_pair")
+load("@rules_python//python:py_runtime.bzl", "py_runtime")
+load("@rules_python//python:py_runtime_pair.bzl", "py_runtime_pair")
 
 cc_library(
    name = "python_headers",
@@ -269,3 +271,23 @@ system_python = repository_rule(
         "minimum_python_version": attr.string(default = "3.9"),
     },
 )
+
+def _system_python_extension(ctx):
+    for mod in ctx.modules:
+        for py in mod.tags.find:
+            system_python(
+                name = py.name,
+                minimum_python_version = py.minimum,
+            )
+
+find = tag_class(attrs = {
+    "name": attr.string(doc = "Supported versions of python to find"),
+    "minimum": attr.string(),
+})
+
+system_python_extension = module_extension(
+    implementation = _system_python_extension,
+    tag_classes = {
+        "find": find,
+    },
+)
